pkg/cmd/experimental/diagnostics/diagnostics.go

```diff
@@ -0,0 +1,214 @@
+
+NOTE: This is an alpha version of diagnostics and will change significantly.
+NOTE: Global flags (from the 'options' subcommand) are ignored here but
+can be used with subcommands.
+`
+
+func NewCommand(name string) *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "diagnostics",
+		Short: "This utility helps you understand and troubleshoot OpenShift v3.",
+		Long:  fmt.Sprintf(longAllDescription, name),
+	}
+	diagFlags := types.NewFlags(cmd.PersistentFlags())
+	addFlags(cmd, diagFlags)
+	cmd.Flags().StringVar(&diagFlags.ClientConfigPath, config.OpenShiftConfigFlagName, "", "Path to the client config file.")
+	cmd.Flags().StringVar(&diagFlags.MasterConfigPath, "master-config", "", "Path to the master config file.")
+	cmd.Flags().StringVar(&diagFlags.NodeConfigPath, "node-config", "", "Path to the node config file.")
+
+	/* There is some weirdness with diagnostics flag usage. The same flags
+	   object is shared between the (implied) "all" cmd and the client cmd.
+
+	   We actually use the client factory built in the "client" subcommand for
+	   discovery in either case, and that adds flags to the "client" cmd which
+	   need a target for putting results in; we do not want the factory to add
+	   those flags to the "all" command (the only client option there is a
+	   config file), but those flags have to exist somewhere for the factory
+	   to set values and look them up later (even though on the "all" command
+	   the flags will not exist).
+
+	   So the flags object from client cmd is reused for the "all" command.
+	*/
+
+	ccmd, factory := NewClientCommand(name+" client", diagFlags)
+	cmd.AddCommand(ccmd)
+	cmd.AddCommand(NewMasterCommand(name + " master"))
+	cmd.AddCommand(NewNodeCommand(name+" node", diagFlags))
+	cmd.AddCommand(NewOptionsCommand())
+
+	cmd.Run = func(c *cobra.Command, args []string) {
+		runInit(c, diagFlags)
+		diagFlags.CanCheck[types.ClientTarget] = true
+		diagFlags.CanCheck[types.MasterTarget] = true
+		diagFlags.CanCheck[types.NodeTarget] = true
+		run.Diagnose(diagFlags, factory)
+	}
+	return cmd
+}
+
+// Used in the command definition for common flags that we want to be visible in usage
+func addFlags(cmd *cobra.Command, flags *types.Flags) {
+	cmd.Flags().VarP(&flags.Diagnostics, "diagnostics", "d", `comma-separated list of diagnostic names to run, e.g. "systemd.AnalyzeLogs"`)
+	cmd.Flags().IntVarP(&flags.LogLevel, "diaglevel", "l", 3, "Level of output: 0: Error, 1: Warn, 2: Notice, 3: Info, 4: Debug")
+	cmd.Flags().StringVarP(&flags.Format, "output", "o", "text", "Output format: text|json|yaml")
+}
+
+// Every command invocation needs to do the same things at the beginning...
+func runInit(cmd *cobra.Command, diagFlags *types.Flags) {
+	cmd.SetOutput(os.Stdout)                         // for output re usage / help
+	diagFlags.OpenshiftFlags = cmd.PersistentFlags() // capture flags from *this* command in env
+	log.SetLevel(diagFlags.LogLevel)
```

Comment:
Log level. Sure it's side-effecty. Logging is side-effecty. Before it has even determined anything there is stuff to output, and it has to know whether it's of sufficient log level to output, and what format to use for the log.

Probably a logger object can be used as you say.